wallet-api/src/Wallet/UTXO/Index.hs

```diff
@@ -0,0 +1,59 @@
+-- | A map of transaction hashes to transactions for efficient lookups of
+--   values on the mockchain.
+module Wallet.UTXO.Index(
+    BlockchainLookup(..),
+    BlockchainIndex(..),
+    empty,
+    insert,
+    initialise,
+    Lookup,
+    runLookup,
+    LookupError(..)
+    ) where
+
+import           Control.Monad.Except (MonadError (..), liftEither)
+import           Control.Monad.Reader (MonadReader (..), ReaderT (..), ask)
+import           Data.Foldable        (fold, foldl')
+import qualified Data.Map             as Map
+import           Data.Semigroup       (Semigroup)
+import           Prelude              hiding (lookup)
+import           Wallet.UTXO.Types    (Blockchain, Tx, TxId', hashTx)
+
+-- | The ability to find transactions by their hashes
+class BlockchainLookup m where
+
+    -- | Find a transaction by its hash
+    lookupTxn :: TxId' -> m Tx
```

Comment:
Worth documenting the lookup approach here: you've chosen to make lookup failure not be handled by the code that does the lookup, it should only look up things it knows are there.

Reply:
Good point, I will document this (see my comment above for explanation)